src/vw/src/main/scala/VowpalWabbitFeaturizer.scala

```diff
@@ -0,0 +1,140 @@
+import org.vowpalwabbit.spark.VowpalWabbitMurmur
+import org.apache.spark.ml.linalg.Vectors
+import org.apache.spark.ml.util.{ComplexParamsReadable, ComplexParamsWritable, Identifiable}
+import org.apache.spark.ml.linalg.SQLDataTypes.VectorType
+
+import scala.collection.mutable.ArrayBuilder
+
+object VowpalWabbitFeaturizer extends ComplexParamsReadable[VowpalWabbitFeaturizer]
+
+class VowpalWabbitFeaturizer(override val uid: String) extends Transformer
+  with HasInputCols with HasOutputCol with HasNumBits with Wrappable with ComplexParamsWritable
+{
+  def this() = this(Identifiable.randomUID("VowpalWabbitFeaturizer"))
+
+  setDefault(inputCols -> Array())
+
+  val seed = new IntParam(this, "seed", "Hash seed")
+  setDefault(seed -> 0)
+
+  def getSeed: Int = $(seed)
+  def setSeed(value: Int): this.type = set(seed, value)
+
+  val stringSplitInputCols = new StringArrayParam(this, "stringSplitInputCols",
+    "Input cols that should be split add word boundaries")
+  setDefault(stringSplitInputCols -> Array())
+
+  def getStringSplitInputCols: Array[String] = $(stringSplitInputCols)
+  def setStringSplitInputCols(value: Array[String]): this.type = set(stringSplitInputCols, value)
+
+  private def getAllInputCols = getInputCols ++ getStringSplitInputCols
+
+  private def getFeaturizer(name: String, dataType: DataType, idx: Int, namespaceHash: Int): Featurizer = {
+    val stringSplitInputCols = getStringSplitInputCols
+
+    dataType match {
+      case DoubleType => new NumericFeaturizer(idx, name, namespaceHash, getMask, r => r.getDouble(idx))
+      case FloatType => new NumericFeaturizer(idx, name, namespaceHash, getMask, r => r.getFloat(idx).toDouble)
+      case IntegerType => new NumericFeaturizer(idx, name, namespaceHash, getMask, r => r.getInt(idx).toDouble)
+      case LongType => new NumericFeaturizer(idx, name, namespaceHash, getMask, r => r.getLong(idx).toDouble)
+      case ShortType => new NumericFeaturizer(idx, name, namespaceHash, getMask, r => r.getShort(idx).toDouble)
+      case ByteType => new NumericFeaturizer(idx, name, namespaceHash, getMask, r => r.getByte(idx).toDouble)
+      case BooleanType => new BooleanFeaturizer(idx, name, namespaceHash, getMask)
+      case StringType =>
+        if (stringSplitInputCols.contains(name))
+          new StringSplitFeaturizer(idx, name, namespaceHash, getMask)
+      else new StringFeaturizer(idx, name, namespaceHash, getMask)
+      case arr: ArrayType => {
+        if (arr.elementType != DataTypes.StringType)
+          throw new RuntimeException(s"Unsupported array element type: ${dataType}")
+
+        new StringArrayFeaturizer(idx, name, namespaceHash, getMask)
+      }
+      case m: MapType => {
+        if (m.keyType != DataTypes.StringType)
+          throw new RuntimeException(s"Unsupported map key type: ${dataType}")
+
+        m.valueType match {
+          case StringType => new MapStringFeaturizer(idx, name, namespaceHash, getMask)
```

Comment:
Seems like this could be cleaned up some using recursion but im not sure what the logic should be so i will defer to your expertise

Reply:
I implemented map of map of map ... thing for the JSON parser in VW - seems the natural thing to do from a dev perspective. 
In hindsight I think it was a mistake as it confuses people more than it helps.